torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1246,6 +1246,174 @@ void LoopNest::computeAt(Stmt* s, For* f) {
+    if (dims[i] != traced_args[i]) {
+      std::cerr << "Cannot rfactor: found unique Var in ReduceOp " << *dims[i]
+                << "\n";
+      return;
+    }
+  }
+
+  std::vector<const Expr*> new_dims = {};
+  Buf* tmp_buf = new Buf(new Var("tmp_buf", kHandle), new_dims);
+  auto old_acc = reduce_op->accumulator();
+  auto old_load = dynamic_cast<Load*>(old_acc.node());
+  auto old_indices = old_load->indices();
+  auto new_indices = old_indices;
+  new_indices.push_back(reduction_var);
+  ExprHandle new_acc = ExprHandle(
+      new Load(reduce_op->body().dtype(), tmp_buf, new_indices, new IntImm(1)));
+  // TODO #1 handle the recursive for loop initializers
+  // TODO #2 get rid of for loop initializers
+  auto old_init_expr = dynamic_cast<Store*>(reduce_op->initializer())->value();
+
+  For* init_for = For::make(
+      VarHandle(target_for->var()),
+      ExprHandle(target_for->start()),
+      ExprHandle(target_for->stop()),
+      new Store(
+          tmp_buf,
+          new_indices,
+          new Cast(reduce_op->body().dtype(), old_init_expr),
+          new IntImm(1)));
+
+  auto new_inner = reduce_op->reduce_args();
+  bool found = false;
+  for (size_t i = 0; i < new_inner.size(); ++i) {
+    if (new_inner[i] == reduction_var) {
+      new_inner.erase(new_inner.begin() + i);
+      found = true;
+      break;
+    }
+  }
+  if (!found) {
+    std::cerr << "Couldn't find target reduction var in the reduce operation\n";
+    return;
+  }
+
+  auto first_reduce = new ReduceOp(
+      new_acc,
+      init_for,
+      reduce_op->body(),
+      reduce_op->interaction(),
+      new_inner);
+
+  auto second_reduce_load = ExprHandle(new Load(
+      reduce_op->body().dtype(), tmp_buf, {reduction_var}, new IntImm(1)));
+  auto second_reduce = new ReduceOp(
+      old_acc,
+      reduce_op->initializer(),
+      second_reduce_load,
+      reduce_op->interaction(),
+      {reduction_var});
+
+  auto root_block = dynamic_cast<Block*>(root_stmt());
+  // 1) replace target for loop (which is a reduction loop)
+  // with an iterative for loop by removing the reduction var from the
+  // innermost op and creating a new temporary output buffer.
+  //
+  // 2) append a clone of the target for loop (which reduces over multiple
+  // variables) with a reduce over only its var by replacing the reduction op
+  // buffer input with the temporary output buffer and removing other reductions
+  // variables.
+  SwapReduce sr(first_reduce);
+  auto res = root_block->replace_stmt(root_for, root_for->accept_mutator(&sr));
+  if (!res) {
+    std::cerr << "Couldn't replace reduce operation with factored version\n";
+    return;
+  };
+
+  auto second_init = dynamic_cast<const Store*>(second_reduce->initializer());
+  if (!second_init) {
+    std::cerr << "Second init should be a store statement\n";
+    return;
+  }
+  auto second_buf = second_init->buf();
+  auto second_indices = second_init->indices();
+  For* new_for = For::make(
```

Comment:
Nit: if you use `new For` instead of `For::make` you won't need all these {Var,Expr}Handle wrappers.